pkg/monitortests/kubeapiserver/auditloganalyzer/audit_log_provider.go

```diff
@@ -0,0 +1,138 @@
+			defer wg.Done()
+			var microBeginning, microEnd *metav1.MicroTime
+			if nil != beginning {
+				micro := metav1.NewMicroTime(*beginning)
+				microBeginning = &micro
+			}
+			if nil != end {
+				micro := metav1.NewMicroTime(*end)
+				microEnd = &micro
+			}
+			err = processNodeEvents(ctx, kubeClient, nodeName, microBeginning, microEnd, ingestors)
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}(ctx, node.Name)
+	}
+	wg.Wait()
+
+	var errs []error
+	for len(errCh) > 0 {
+		err := <-errCh
+		errs = append(errs, err)
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func processNodeEvents(ctx context.Context, client kubernetes.Interface, nodeName string, beginning, end *v1.MicroTime, ingestors IngestorFactory) error {
+	// we only care to process the kube-apiserver audit logs
+	return processAPIServerEvents(ctx, client, nodeName, "kube-apiserver", beginning, end, ingestors)
+}
+
+func processAPIServerEvents(ctx context.Context, client kubernetes.Interface, nodeName, apiserver string, beginning, end *v1.MicroTime, ingestors IngestorFactory) error {
+	auditLogFilenames, err := getAuditLogFilenames(ctx, client, nodeName, apiserver)
+	if err != nil {
+		return err
+	}
+
+	// we do not have enough memory to read all the content and then navigate it all in memory.
+	// I mean, maybe we do, but we are looking at a number of bytes read measured in gigs, then the deserialization buffers,
+	// and then the maps to actually store the content.
+	// Instead, we will stream, deserialize as we consume, and run a summarize function to get aggregate values.
+	wg := sync.WaitGroup{}
+
+	errCh := make(chan error, len(auditLogFilenames))
+	for _, auditLogFilename := range auditLogFilenames {
+		if !strings.HasPrefix(auditLogFilename, "audit") {
+			continue
+		}
+		if !strings.HasSuffix(auditLogFilename, ".log") {
+			continue
+		}
+		klog.InfoS("processing audit log events", "node", nodeName, "apiserver", apiserver, "filename", auditLogFilename)
+		wg.Add(1)
+		go func(ctx context.Context, auditLogFilename string) {
+			defer wg.Done()
+
+			auditStream, err := nodeaccess.StreamNodeLogFile(ctx, client, nodeName, filepath.Join(apiserver, auditLogFilename))
+			if err != nil {
+				errCh <- err
+				return
+			}
+			ingestor := ingestors.New()
+			scanner := bufio.NewScanner(auditStream)
+			line := 0
+			for scanner.Scan() {
+				line++
+				auditLine := scanner.Bytes()
+
+				if len(auditLine) == 0 {
+					continue
+				}
+
+				auditEvent := &auditv1.Event{}
+				if err := json.Unmarshal(auditLine, auditEvent); err != nil {
+					// TODO
+					// auditLogSummary.lineReadFailureCount++
+					fmt.Printf("unable to decode %q line %d: %s to audit event: %v\n", auditLogFilename, line, string(auditLine), err)
+					continue
+				}
+
+				if beginning != nil && auditEvent.RequestReceivedTimestamp.Before(beginning) || end != nil && end.Before(&auditEvent.RequestReceivedTimestamp) {
+					continue
+				}
+				ingestor.Ingest(auditEvent)
```

Comment:
I think this could be the place where I could register my hook.
Then, I could ingest every audit event and analyze them for high latency